x/mongo/driver/operation.go

```diff
@@ -531,15 +531,17 @@ func (op Operation) Execute(ctx context.Context, scratch []byte) error {
 			serviceID:    startedInfo.serviceID,
 		}
 
-		// Check if there's enough time to perform a round trip before the Context deadline. If ctx is
-		// a Timeout Context, use the 90th percentile RTT as a threshold. Otherwise, use the minimum observed
-		// RTT.
-		if deadline, ok := ctx.Deadline(); ok {
+		// Check for possible context error. If no context error, check if there's enough time to perform a
+		// round trip before the Context deadline. If ctx is a Timeout Context, use the 90th percentile RTT
+		// as a threshold. Otherwise, use the minimum observed RTT.
+		if ctx.Err() != nil {
+			err = ctx.Err()
+		} else if deadline, ok := ctx.Deadline(); ok {
 			if internal.IsTimeoutContext(ctx) && time.Now().Add(srvr.RTTMonitor().P90()).After(deadline) {
 				err = internal.WrapErrorf(ErrDeadlineWouldBeExceeded,
 					"remaining time %v until context deadline is less than 90th percentile RTT\n%v", time.Until(deadline), srvr.RTTMonitor().Stats())
 			} else if time.Now().Add(srvr.RTTMonitor().Min()).After(deadline) {
-				err = op.networkError(context.DeadlineExceeded)
+				err = context.DeadlineExceeded
 			}
 		}
```

Comment:
It's possible for the context to expire after this check and before `writeWireMessage` or `readWireMessage`, in which case an error will be returned from the connection level and may still be marked as retryable because of `networkError()`. It's difficult, though, to disambiguate between a context expiration from the connection level and a socket timeout. And, once we retry the operation, the expired context will be picked up in the block above and _not_ marked as retryable, so we still will not retry indefinitely with expired contexts.